src/imagefilters.cpp

```diff
@@ -0,0 +1,173 @@
+		u32 ss = 0, sr = 0, sg = 0, sb = 0;
+
+		// Walk each neighbor pixel (clipped to image bounds).
+		for (u32 sy = (ctry < 1) ? 0 : (ctry - 1);
+				sy <= (ctry + 1) && sy < dim.Height; sy++)
+		for (u32 sx = (ctrx < 1) ? 0 : (ctrx - 1);
+				sx <= (ctrx + 1) && sx < dim.Width; sx++) {
+
+			// Ignore transparent pixels.
+			irr::video::SColor d = src->getPixel(sx, sy);
+			if (d.getAlpha() <= threshold)
+				continue;
+
+			// Add RGB values weighted by alpha.
+			u32 a = d.getAlpha();
+			ss += a;
+			sr += a * d.getRed();
+			sg += a * d.getGreen();
+			sb += a * d.getBlue();
+		}
+
+		// If we found any neighbor RGB data, set pixel to average
+		// weighted by alpha.
+		if (ss > 0) {
+			c.setRed(sr / ss);
+			c.setGreen(sg / ss);
+			c.setBlue(sb / ss);
+			src->setPixel(ctrx, ctry, c);
+		}
+	}
+}
+
+/* Scale a region of an image into another image, using nearest-neighbor with
+ * anti-aliasing; treat pixels as crisp rectangles, but blend them at boundaries
+ * to prevent non-integer scaling ratio artifacts.  Note that this may cause
+ * some blending at the edges where pixels don't line up perfectly, but this
+ * filter is designed to produce the most accurate results for both upscaling
+ * and downscaling.
+ */
+void imageScaleNNAA(video::IImage *src, const core::rect<s32> &srcrect, video::IImage *dest) {
+
+	double minsx, maxsx, minsy, maxsy, area, ra, ga, ba, aa, pw, ph, pa, tmp;
+	u32 dy, dx;
+	s32 sy, sx, maxs32sx, maxs32sy;
+	video::SColor pxl;
+
+	// Cache rectsngle boundaries.
+	double sox = srcrect.UpperLeftCorner.X * 1.0;
+	double soy = srcrect.UpperLeftCorner.Y * 1.0;
+	double sw = srcrect.getWidth() * 1.0;
+	double sh = srcrect.getHeight() * 1.0;
+
+	// Walk each destination image pixel.
+	// Note: loop y around x for better cache locality.
+	core::dimension2d<u32> dim = dest->getDimension();
+	for (dy = 0; dy < dim.Height; dy++)
+	for (dx = 0; dx < dim.Width; dx++) {
+
+		// Calculate floating-point source rectangle bounds.
+		// Do some basic clipping, and for mirrored/flipped rects,
+		// make sure min/max are in the right order.
+		minsx = sox + (dx * sw / dim.Width);
+		if (minsx < 0)
+			minsx = 0;
+		else if (minsx > sw)
+			minsx = sw;
+		maxsx = minsx + sw / dim.Width;
+		if (maxsx < 0)
+			maxsx = 0;
+		else if (maxsx > sw)
+			maxsx = sw;
+		if (minsx > maxsx) {
+			tmp = minsx;
+			minsx = maxsx;
+			maxsx = tmp;
+		}
+		minsy = soy + (dy * sh / dim.Height);
+		if (minsy < 0)
+			minsy = 0;
+		else if (minsy > sh)
+			minsy = sh;
+		maxsy = minsy + sh / dim.Height;
+		if (maxsy < 0)
+			maxsy = 0;
+		else if (maxsy > sh)
+			maxsy = sh;
+		if (minsy > maxsy) {
+			tmp = minsy;
+			minsy = maxsy;
+			maxsy = tmp;
+		}
+
+		// Total area, and integral of r, g, b values over that area,
+		// initialized to zero, to be summed up in next loops.
+		area = 0;
+		ra = 0;
+		ga = 0;
+		ba = 0;
+		aa = 0;
+
+		// Loop over the integral pixel positions described by those bounds.
+		maxs32sx = (s32)maxsx;
```

Comment:
The question that pops to my mind is "why is 'sy'" signed? At this point in the code, minsy and minsx are both >= 0 therefore sy and sx can both be u32 and all the casts are not necessary.

(Incidentally, minsy and minsx are already s32 so I'm not sure why you're casting them on lines 132 and 133 :))